README.md
sip-proxpy
==========

The proxpy Python proxy as used in DAVID Social Innovation Platform development


CHANGES
Change history
---------------
[1.2]
    - Made installable package and tested installation and import of proxpy
      modules in other applications
    - Added plugin for the shiny-httpheader-test example of rewriting shiny 
      server pro http headers
[1.1] 
    - Added fixes from https://github.com/yeukhon/proxpy/
[1.0] 
    - start based on r29 of https://code.google.com/p/proxpy/




sipproxpy/https.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
  Copyright notice
  ================
  
  Copyright (C) 2011
      Roberto Paleari     <roberto.paleari@gmail.com>
      Alessandro Reina    <alessandro.reina@gmail.com>
  
  This program is free software: you can redistribute it and/or modify it under
  the terms of the GNU General Public License as published by the Free Software
  Foundation, either version 3 of the License, or (at your option) any later
  version.
  
  HyperDbg is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
  A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License along with
  this program. If not, see <http://www.gnu.org/licenses/>.
  
"""

from http import *
import select

class HTTPSRequest(HTTPMessage):
    @staticmethod
    def sendAck(socket):
        # Send a 200 response to acknowledge the connection
        ack = HTTPResponse("HTTP/1.1", HTTPMessage.HTTP_CODE_OK, "Connection Established")
        socket.send(ack.serialize())

class HTTPSUtil():
    @staticmethod
    def wait_read(socket):
        if socket.pending():
            return
        select.select([socket], [], [])


sipproxpy/logger.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
  Copyright notice
  ================
  
  Copyright (C) 2011
      Roberto Paleari     <roberto.paleari@gmail.com>
      Alessandro Reina    <alessandro.reina@gmail.com>
  
  This program is free software: you can redistribute it and/or modify it under
  the terms of the GNU General Public License as published by the Free Software
  Foundation, either version 3 of the License, or (at your option) any later
  version.
  
  HyperDbg is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
  A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License along with
  this program. If not, see <http://www.gnu.org/licenses/>.
  
"""

import os, sys
import threading

COLOR_RED    = 31
COLOR_GREEN  = 32
COLOR_YELLOW = 33
COLOR_BLUE   = 34
COLOR_PURPLE = 35

def colorize(s, color = COLOR_RED):
    return (chr(0x1B) + "[0;%dm" % color + str(s) + chr(0x1B) + "[0m")

class Logger:
    def __init__(self, verbosity = 0):
        self.verbosity = verbosity

    def __out(self, msg, head, color):
        tid = threading.current_thread().ident & 0xffffffff
        tid = " %s " % colorize("<%.8x>" % tid, COLOR_PURPLE)
        print colorize(head, color) + tid + msg    

    def info(self, msg):
        self.__out(msg, "[*]", COLOR_GREEN)

    def warning(self, msg):
        self.__out(msg, "[#]", COLOR_YELLOW)

    def error(self, msg):
        self.__out(msg, "[!]", COLOR_RED)

    def debug(self, msg):
        if self.verbosity > 0:
            self.__out(msg, "[D]", COLOR_BLUE)


sipproxpy/proxpy.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
  Copyright notice
  ================
  
  Copyright (C) 2011
      Roberto Paleari     <roberto.paleari@gmail.com>
      Alessandro Reina    <alessandro.reina@gmail.com>
  
  This program is free software: you can redistribute it and/or modify it under
  the terms of the GNU General Public License as published by the Free Software
  Foundation, either version 3 of the License, or (at your option) any later
  version.
  
  HyperDbg is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
  A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License along with
  this program. If not, see <http://www.gnu.org/licenses/>.
  
"""

import sys
import getopt

from core import *

def show_help():
    print """\
Syntax: python %s <options>
 -a <addr>         listen address (default 0.0.0.0)
 -d <filename>     on termination, dump requests & responses to file
 -h                show this help screen
 -p <port>         listen port  (default 8080)
 -r <host:[port]>  redirect HTTP traffic to target host (default port: 80)
 -v                be more verbose
 -x <filename>     load a ProxPy plugin
""" % sys.argv[0]

def parse_options():
    try:
        opts, args = getopt.getopt(sys.argv[1:], "a:d:hp:r:vx:")
    except getopt.GetoptError, e:
        print str(e)
        show_help()
        exit(1)

    opts = dict([(k.lstrip('-'), v) for (k,v) in opts])

    if 'h' in opts:
        show_help()
        exit(0)

    ps = ProxyState()

    if 'v' in opts:
        ps.log.verbosity += 1

    if 'd' in opts:
        ps.dumpfile = opts['d']

    if 'p' in opts:
        ps.listenport = int(opts['p'])
        
    if 'a' in opts:
        ps.listenaddr = opts['a']

    # Check and parse redirection host
    if 'r' in opts:
        h = opts['r']
        if ':' not in h:
            p = 80
        else:
            h,p = h.split(':')
            p = int(p)
        ps.redirect = (h, p)

    # Load an external plugin
    if 'x' in opts:
        ps.plugin = ProxyPlugin(opts['x'])

    return ps

def main():
    global proxystate
    try:
        proxystate = parse_options()
        proxyServer = ProxyServer(proxystate)
        proxyServer.startProxyServer()
    except KeyboardInterrupt, e:
        nreq, nres = proxystate.history.count()
        proxystate.log.info("Terminating... [%d requests, %d responses]" % (nreq, nres))
        if proxystate.dumpfile is not None:
            data = proxystate.history.dumpXML()
            f = open(proxystate.dumpfile, 'w')
            f.write(data)
            f.close()

if __name__ == "__main__":
    global proxystate
    main()


sipproxpy/history.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
  Copyright notice
  ================
  
  Copyright (C) 2011
      Roberto Paleari     <roberto.paleari@gmail.com>
      Alessandro Reina    <alessandro.reina@gmail.com>
  
  This program is free software: you can redistribute it and/or modify it under
  the terms of the GNU General Public License as published by the Free Software
  Foundation, either version 3 of the License, or (at your option) any later
  version.
  
  HyperDbg is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
  A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License along with
  this program. If not, see <http://www.gnu.org/licenses/>.
  
"""

import threading
import datetime, base64
from xml.sax.saxutils import escape as xmlescape

# Synchronization decorator
def synchronized(lock):
    def wrap(f):
        def new_function(*args, **kw):
            lock.acquire()
            try:
                return f(*args, **kw)
            finally:
                lock.release()
        return new_function
    return wrap

class HttpHistoryEntry:
    def __init__(self, idz, oreq = None, mreq = None, ores = None, mres = None):
        self.id   = idz         # Entry identified (mandatory)
        self.setOriginalRequest(oreq)
        self.setOriginalResponse(ores)
        self.setMangledRequest(mreq)
        self.setMangledResponse(mres)

    def setOriginalRequest(self, r):
        if r is None:
            t = None
        else:
            t = datetime.datetime.now()
        self.oreq_time = t
        self.oreq = r

    def setOriginalResponse(self, r):
        if r is None:
            t = None
        else:
            t = datetime.datetime.now()
        self.ores_time = t
        self.ores = r

    def setMangledRequest(self, r):
        if r is None:
            t = None
        else:
            t = datetime.datetime.now()
        self.mreq_time = t
        self.mreq = r

    def setMangledResponse(self, r):
        if r is None:
            t = None
        else:
            t = datetime.datetime.now()
        self.mres_time = t
        self.mres = r

class HttpHistory:
    # Synchronization lock
    lock  = threading.Lock()

    def __init__(self):
        self.__history = []

    @synchronized(lock)
    def allocate(self):
        idz = len(self.__history)
        h = HttpHistoryEntry(idz = idz)
        self.__history.append(h)
        return idz

    @synchronized(lock)
    def __getitem__(self, idz):
        return self.__history[idz]

    def count(self):
        """
        Count requests and responses. Return a tuple (#req, #res).
        """
        nreq, nres = 0, 0
        for entry in self.__history:
            if entry.oreq is not None:
                nreq += 1
            if entry.ores is not None:
                nres += 1
        return nreq, nres

    def dumpXML(self):
        t = datetime.datetime.now()

        # Document header
        s = """\
<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\" ?>
<Entries>
"""

        # Process single HTTP entries
        for entry in self.__history:
            s += "  <Entry>\n"
            s += "    <ID>%d</ID>\n" % entry.id

            for attr, name in [
                ("oreq", "OriginalRequest"),
                ("mreq", "MangledRequest"),
                ("ores", "OriginalResponse"),
                ("mres", "MangledResponse"),
                ]:

                v = getattr(entry, attr)
                t = getattr(entry, attr + "_time")
                if v is not None:
                    s += """\
    <%s>
      <Timestamp>%s</Timestamp>
      <Data>
""" % (name, t)
                    proto = getattr(v, "proto")
                    code = hasattr(v, "code")
                    msg = hasattr(v, "msg")
                    if code and msg:
                        s += """\
        <Protocol>%s</Protocol>
        <StatusCode>%s</StatusCode>
        <StatusMsg>%s</StatusMsg>
""" % (proto, v.code, v.msg)
                    else:
                        s += """\
        <Protocol>%s</Protocol>
        <Method>%s</Method>
""" % (proto, v.method)
                    print v.headers_as_str()
                    # Process entry headers
                    for hname, hvalues in v.headers.iteritems():
                        for hvalue in hvalues:
                            s += """\
          <Header>
            <Name>%s</Name>
            <Value>%s</Value>
          </Header>
""" % (xmlescape(hname), xmlescape(hvalue))

                    # Append raw headers
                    s += """\
          <RawHeaders>%s</RawHeaders>
""" % v.headers_as_str()

                    if (hasattr(v, "getHost")):
                        # Append the scheme, host and port
                        s += """\
              <Url>
                <Scheme>%s</Scheme>
                <HostName>%s</HostName>
                <Port>%s</Port>
                <Path>%s</Path>
              </Url>
    """ % ("http", v.getHost()[0], v.getHost()[1], v.getPath())
                    # Process entry body and close tag
                    s += """\
        <Body>%s</Body>
      </Data>
    </%s>
""" % (base64.encodestring(v.body), name)

            s += "  </Entry>\n"

        s += "</Entries>\n"

        return s



sipproxpy/http.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
  Copyright notice
  ================

  |Copyright (C) 2011
  |    Roberto Paleari     <roberto.paleari@gmail.com>
  |    Alessandro Reina    <alessandro.reina@gmail.com>

  This program is free software: you can redistribute it and/or modify it under
  the terms of the GNU General Public License as published by the Free Software
  Foundation, either version 3 of the License, or (at your option) any later
  version.

  HyperDbg is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
  A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

  You should have received a copy of the GNU General Public License along with
  this program. If not, see <http://www.gnu.org/licenses/>.

"""

import datetime
import copy
import urlparse
import select

class HTTPUtil():
    @staticmethod
    def wait_read(socket):
        select.select([socket], [], [])

class HTTPMessage():
    EOL = "\r\n"
    HTTP_CODE_OK = 200

    # Global message ID, incremented at each HTTP request/reponse
    uid = 0

    def __init__(self, headers = None):
        self.peer = None
        self.time = datetime.datetime.now()
        # Set unique message ID
        self.uid  = HTTPMessage.uid

        if headers is None:
            self.headers = {}
        elif isinstance(headers, list):
            self.headers = HTTPMessage._readheaders(headers)
        else:
            assert isinstance(headers, dict)
            self.headers = headers

        HTTPMessage.uid += 1

    @staticmethod
    def _readheaders(data):
        headers = {}

        for line in data:
            if line == HTTPMessage.EOL:
                break
            assert ":" in line
            line = line.rstrip(HTTPMessage.EOL)
            i = line.index(":")
            n = line[:i]
            v = line[i+1:]
            if n not in headers:
                headers[n] = []
            headers[n].append(v.lstrip())

        return headers

    @staticmethod
    def _readbody(data, headers):
        bodylen = None
        chunked = False
        for n,v in headers.iteritems():
            if n.lower() == "content-length":
                assert bodylen is None, "[!] Duplicated content length"
                bodylen = int(v[0])
            elif n.lower() == "transfer-encoding" and v[0].lower() == "chunked":
                chunked = True
                break

        # Read HTTP body (if present)
        body = ""
        if bodylen is not None:
            body = data.read(bodylen)
        elif chunked:
            # Chunked encoding
            while True:
                # Determine chunk length
                chunklen = data.readline()
                chunklen = int(chunklen, 16)

                # Read the whole chunk
                chunk = ""
                chunk = data.read(chunklen)
                body += chunk

                if chunklen == 0: 
                    break

                # Read trailing CRLF
                eol = data.read(2)
                assert eol == HTTPMessage.EOL

        return body

    def isChunked(self):
        r = False
        for n, v in self.headers.iteritems():
            if n.lower() == "transfer-encoding" and v[0].lower() == "chunked":
                r = True
                break
        return r

    def isKeepAlive(self):
        if 'Connection' in self.headers:
            if self.headers['Connection'][0] == 'keep-alive':
                return True
        elif 'Proxy-Connection' in self.headers:
            if self.headers['Proxy-Connection'][0] == 'keep-alive':
                return True

        return False

    def setPeer(self, h, link = True):
        self.peer = h
        if link:
            h.setPeer(self, link = False)

    def clone(self):
        return copy.deepcopy(self)

    def fixup(self):
        # Fix headers
        for n in self.headers:
            if n.lower() == "content-length":
                self.headers[n][0] = len(self.body)

    # Hack to fix HTTPS request 
    @staticmethod
    def _fixURLMalformed(scheme, url, headers):
        if ((url.find('http') != 0) and (url[0] == '/')):
            assert 'Host' in headers
            url = scheme + '://' + headers['Host'][0] + url
        return url

    def __findHeader(self, name, ignorecase = True):
        r = None
        for n in self.headers:
            if (ignorecase and name.lower() == n.lower()) or ((not ignorecase) and name == n):
                r = n
                break
        return r

    def getHeader(self, name, ignorecase = True):
        """
        Get the values of header(s) with name 'name'. If 'ignorecase' is True,
        then the case of the header name is ignored.
        """
        r = []
        for n,v in self.headers.iteritems():
            if (ignorecase and name.lower() == n.lower()) or ((not ignorecase) and name == n):
                r.extend(v)
        return r

    def addHeader(self, name, value, ignorecase = True):
        """
        Add a new 'name' header with value 'value' to this HTTPMessage. If
        'ignorecase' is True, then the case of the header name is ignored.
        """
        k = self.__findHeader(name, ignorecase)
        if k not in self.headers:
            self.headers[name] = []
        self.headers[name].append(value)

    def setHeader(self, name, value, ignorecase = True):
        """
        Set header with name 'name' to 'value'. Any existing header with the
        same name is replaced. If 'ignorecase' is True, then the case of the
        header name is ignored.
        """
        k = self.__findHeader(name, ignorecase)
        self.headers[k] = [value, ]


class HTTPRequest(HTTPMessage):
    METHOD_GET     = 1
    METHOD_POST    = 2
    METHOD_HEAD    = 3
    METHOD_OPTIONS = 4
    METHOD_CONNECT = 5

    def __init__(self, method, url, proto, headers = None, body = ""):
        self.method  = method
        self.url     = url
        self.proto   = proto
        self.body    = body
        HTTPMessage.__init__(self, headers)

    @staticmethod
    def build(data):
        # Read request line
        reqline = data.readline().rstrip(HTTPMessage.EOL)

        if reqline == '': 
            return None

        method, url, proto = reqline.split()

        # Read headers & body
        headers = HTTPMessage._readheaders(data)
        body    = HTTPMessage._readbody(data, headers)
        url = HTTPMessage._fixURLMalformed("http", url, headers)
        return HTTPRequest(method, url, proto, headers, body)

    def getHost(self):
        if self.getMethod() == HTTPRequest.METHOD_CONNECT:
            tmp = self.url.split(":")
            host = tmp[0]
            if len(tmp) > 0:
                port = int(tmp[1])
            else:
                port = 80
        else:
            r = urlparse.urlparse(self.url)
            port = r.port
            if port is None and r.scheme != "https":
                port = 80
            elif r.scheme == "https":
                port = 443
            host = r.hostname

        assert host is not None and len(host) > 0, "[!] Cannot find target host in URL '%s'" % self.url
        return host, port

    def getPath(self):
        r = urlparse.urlparse(self.url)
        s = r.path
        if len(r.params) > 0:
            s += ";%s" % r.params
        if len(r.query) > 0:
            s += "?%s" % r.query
        if len(r.fragment) > 0:
            s += "#%s" % r.fragment
        return s

    def headers_as_str(self):
        s = ""
        for n,v in self.headers.iteritems():
            print n
            if n not in ("Host", "User-Agent"):
                for i in v:
                    s += "%s: %s\n" % (n, i)
        return s

    def __str__(self):
        s = "%s %s %s\n" % (self.method, self.getPath(), self.proto)
        print self.headers.iteritems()
        for n,v in self.headers.iteritems():
            for i in v:
                s += "%s: %s\n" % (n, i)
        return s

    def isRequest(self):
        return True

    def getMethod(self):
        m = self.method.lower()
        if   m == "get":     r = HTTPRequest.METHOD_GET
        elif m == "post":    r = HTTPRequest.METHOD_POST
        elif m == "head":    r = HTTPRequest.METHOD_HEAD
        elif m == "options": r = HTTPRequest.METHOD_OPTIONS
        elif m == "connect": r = HTTPRequest.METHOD_CONNECT
        elif m == "unknown": r = HTTPRequest.METHOD_UNKNOWN
        return r

    def getParams(self, typez = None):
        params = {}
        if typez is None or typez == HTTPRequest.METHOD_GET:
            r = urlparse.urlparse(self.url).query
            if len(r) > 0:
                params.update(urlparse.parse_qs(r))
        if typez is None or typez == HTTPRequest.METHOD_POST:
            if len(self.body) > 0:
                params.update(urlparse.parse_qs(self.body, keep_blank_values = True))

        if params:
            # FIXME: Do we lose v[1:] ?
            tmp = {}
            for k, v in params.iteritems():
                tmp[k] = v[0]
            params = tmp

        return params

class HTTPResponse(HTTPMessage):
    def __init__(self, proto, code, msg, headers = None, body = ""):
        self.proto   = proto
        self.code    = code
        self.msg     = msg
        self.body    = body
        HTTPMessage.__init__(self, headers)

    @staticmethod
    def build(data):
        # Read request line
        reqline = data.readline().rstrip(HTTPMessage.EOL)

        method, url, proto = reqline.split()

        # Read headers & body
        headers = HTTPMessage._readheaders(data)
        body    = HTTPMessage._readbody(data, headers)

        return HTTPRequest(method, url, proto, headers, body)

    def serialize(self):
        # Response line
        s = "%s %s %s" % (self.proto, self.code, self.msg)
        s += HTTPMessage.EOL

        # Headers
        for n,v in self.headers.iteritems():
            for i in v:
                s += "%s: %s" % (n, i)
                s += HTTPMessage.EOL

        s += HTTPMessage.EOL

        # Body
        if not self.isChunked():
            s += self.body
        else:
            # FIXME: Make a single-chunk body
            s += "%x" % len(self.body) + HTTPMessage.EOL
            s += self.body + HTTPMessage.EOL
            s += HTTPMessage.EOL
            s += "0" + HTTPMessage.EOL + HTTPMessage.EOL

        return s

    def headers_as_str(self):
        s = "%s %s %s\n" % (self.proto, self.code, self.msg)
        for n,v in self.headers.iteritems():
            for i in v:
                s += "%s: %s\n" % (n, i)
        return s

    def __str__(self):
        s = "%s %s %s\n" % (self.proto, self.code, self.msg)
        for n,v in self.headers.iteritems():
            for i in v:
                s += "%s: %s\n" % (n, i)
        s = s + "\n" + self.body
        return s

    def isResponse(self):
        return True